### ISSUE TITLE
jack / jack_transport issue (crash)

### ISSUE THREAD

**Issue by grammoboy2** [M0JXD/non-sequencer-fork#1]
FYI I'm using non-sequencer from arch linux and it crashed. I got the feeling it has something to do with jack_transport and positioning. 
https://aur.archlinux.org/packages/non-sequencer

**Comment by grammoboy2** [comment]
didn't @theGreatWhiteShark added some jack transport changes to non-daw?

**Comment by M0JXD** [comment]
I haven't had any jack transport related crashes but I've yet to give everything a proper going through and test. 

How old is the source for the arch package? The stated upstream has been defunct for some time, from what I can tell it's being built from https://github.com/linuxaudio/non.

That code may have already been fixed before I even forked it. Certainly at least we've found out that the KXStudio binary packages are nearly a year older than the @original-male 's last commit, and that is version 1.20, the Arch one says 1.10. I'll certainly have a look at the work done by @theGreatWhiteShark though, thanks!

**Comment by theGreatWhiteShark** [comment]
Hmmm. I never really used the sequencer part of Non. I do recall fixing a JACK synchronization (timebase stuff, I think) issue in Non Timeline and a UI resize issue in Non Session Manager. But apart from that I didn't do much in the Non code base.

**Comment by M0JXD** [comment]
Thank you for the details - I haven't had to touch the JACK code yet, a lot of it seems to be handled by the shared code in nonlib anyways which I've pulled in from Stazed, I'll find your old commit and see if sequencer might be affected by the same problem.

**Comment by grammoboy2** [comment]
`*** buffer overflow detected ***: terminated`

JACK: 
```
JackEngine::XRun: client = Non-Sequencer.nUMAD was not finished, state = Triggered
JackAudioDriver::ProcessGraphAsyncMaster: Process error
ClientNotify fails name = Carla-Multi-Client.nVWQK notification = 18 val1 = 1 val2 = 0
JackEngine::XRun: client = Non-Sequencer.nUMAD was not finished, state = Triggered
Cannot write socket fd = 86 err = Broken pipe
CheckRes error
Could not write notification
ClientNotify fails name = Non-Sequencer.nUMAD notification = 18 val1 = 0 val2 = 0
Cannot write socket fd = 86 err = Broken pipe
CheckRes error
Could not write notification
ClientNotify fails name = Non-Sequencer.nUMAD notification = 18 val1 = 1 val2 = 0
Unknown request 4294967295
CheckSize error size = 0 Size() = 12
CheckRead error
```

**Comment by M0JXD** [comment]
Is this still the Arch package or compiled from my fork?

**Comment by M0JXD** [comment]
Is this still the arch package or compiled from my source?

**Comment by M0JXD** [comment]
Are you using under a SM? Looking here there was an problem when NSM was first introduced that caused JACK issues when used without a SM, but was promptly fixed:

https://github.com/falkTX/non/commits/a320232b66363efa531187632d3907865a41c099/src/jack.C?browsing_rename_history=true&new_path=sequencer/src/jack.C&original_branch=master

I've had both branches of my fork working great as transport slaves (the one without optional gui is still using the old "nonlib" too) running some patterns in a session with Non-Timeline-XT as the transport master, I can move the playhead however I wish during playback and it syncs up perfectly with no xruns/crashes.

However it seems that if sequencer is the timebase master (which will never be so when used in a session with Non-Timeline) their are a few different errors that can occur. 

When an external source (e.g. QjackCtl) tries to rewind/fast-forward JACK transport while paused (e.g. to set to 0 and then play from start) the GUI crashes but the audio engine keeps running. It also becomes unresponsive to NSM (which is no surprise as that runs in the GUI thread.)

If there is a (QjackCtl) request to rewind whilst running the you will get a JACK notification (not really a full xrun) that "Non-Sequencer was not finished, state = Running" but no crash. Non-Mixer gave similar errors too, I think this is expected behavior if Non-Sequencer is the timebase master.

I'll look some more in to this.

**Comment by grammoboy2** [comment]
I don't see where your version differs from the one of  Arch Linux related to this issue. Yours doesn't build here (and I'm not sure if it's a good idea to add another build system? Zig build would be nice maybe.)
```
[ 35%]: cache compiling.release src/gui/ui.C
error: /usr/include/sigc++-2.0/sigc++/signal_base.h:24:10: fatal error: sigc++config.h: No such file or directory
   24 | #include <sigc++config.h>
      |          ^~~~~~~~~~~~~~~~
compilation terminated.
  > in src/sequence.C
```
Dependency is installed, as it is the same for the arch package. 

Non-sequencer get this error after trying to load any project file, in or outside NSM. 

```
*** buffer overflow detected ***: terminated
Aborted (core dumped)
```

**Comment by M0JXD** [comment]
Ugh, I had trouble getting the build system to detect sigc++ on mine too (the intellisense in my code editor couldn't find it either!). I think it's because both v2 and v3 are available. I saw in the [old github issues](https://web.archive.org/web/20201123000258/https://github.com/original-male/non/issues/210) that it's one of the dependencies JML wanted to remove because it was causing build issues even then.

Could you search and let me know where sigc++config.h is on your system? (Saves me having to spin up an Arch machine in a VM) On mine (Mint) it was "/usr/lib/x86_64-linux-gnu/sigc++-2.0/include"

Regarding the build system, I'm not adding one, I'm replacing it. I like what Xmake offers. The only real competitors are CMake and Meson as far as I see, and even then to match the features you need a package manager like vcpkg/conan, which Xmake can link to automatically. I could see Waf becoming a maintenance difficulty. For me at least when it comes to scripting-esque activities (which is what a build system essentially is, it automates calling the compiler/linker etc.) Lua is preferable to Python, which is preferable to a DSL like Meson or CMake (this one especially does not spark joy). The only other way I'd consider would be an old school Makefile but that comes with it's range of difficulties.

The Zig build system requires you write Zig... I don't know it and I don't fancy trying to use the language intended to be a C replacement for my build system, it's the wrong tool IMO. It will also always be developed with a Zig first mindset. If it suffers an edge case/break for C/C++ building are we sure that will be fixed in a timely manner? 

As far as I see it, Lua is great for this because it's simple but still really fast, unlike Python based ones like Waf. Xmake even beats ninja in some tests when used as a direct building tool.

I'm not saying it's perfect, no build system ever will be, but it's whole lot more preferable for me at least.

**Comment by M0JXD** [comment]
I've just pushed a change to the sigc++ detection which should be more platform agnostic (both branches), try it now.

**Comment by theGreatWhiteShark** [comment]
> Thank you for the details - I haven't had to touch the JACK code yet, a lot of it seems to be handled by the shared code in nonlib anyways which I've pulled in from Stazed, I'll find your old commit and see if sequencer might be affected by the same problem.

IIRC the JACK fixes are only present in my [fork of non](https://github.com/theGreatWhiteShark/non) as I did them shortly after Jonathan went dark.

> *** buffer overflow detected ***: terminated
> 
> JACK:
> ...

The JACK log looks fine. It only tells Non-Sequencer did die within the process cycle. This does not necessarily indicate it's death was JACK-related.

**Comment by grammoboy2** [comment]
I could build it now, after editing https://github.com/M0JXD/non-sequencer-fork/blob/master/xmake.lua#L21 to /usr/ 

No crash. Old skool transparant theme. So far so good.  

Did you (try to) make sure all the wscript (performance) build settings where transferred to xmake?

The biggest challenge would be to maintain the same clean approach and style as JML had (if that's one of the project goals, you seems to be interested in it's origins and philosophy at least, that's something I like).

**Comment by M0JXD** [comment]
Yes, there's a build option there for SSE2 optimisations (enabled by default). I need to double check it though as it relies on Xmake's default release build being without vector extensions, and I don't fully believe that's true (hence it's not mentioned in the README yet). Likewise as said I'll add build options for hide/show GUI, and anything else desired.

I'll try and fix the NTK detection (that's what you had to adjust right? sigc++ worked with my commit?) in the same way as it's done for sigc++.

> IIRC the JACK fixes are only present in my [fork of non](https://github.com/theGreatWhiteShark/non) as I did them shortly after Jonathan went dark.

Thank you for the link! These will likely be very useful as the bugs only occur in timebase master mode, same as Non-Timeline always behaves.

> if that's one of the project goals, you seems to be interested in it's origins and philosophy at least, that's something I like

Just a reminder that while this has some (well, one) new features I still currently intend this to be closer to "life support/maintenance" than an actively developed project. I don't have any ideas of radically changing the program, it's more-so about a few QOL updates and bugfixes as listed in the README. 

I think what I've decided upon is keeping Non-Sequencer as a pattern centric tool for the likes of percussion instruments and parts that it would suit, (hosting plugins in Non-Mixer-XT and/or Carla) and then using something else for more melodical/chordal work - still testing all the options on that, but as mentioned Helio is my favorite so far, although I'm thinking the way it freely creates space when scrolling would not play well with JACK support. Maybe MIDI tracks in Timeline is the way...

I'm open to suggestions though.

**Comment by M0JXD** [comment]
> I'll try and fix the NTK detection

This should be fixed too now.
